5/README.md

```diff
@@ -45,23 +54,138 @@ There are many use cases for the usage of a Property:
 
+* To configure some alarms on Attribute value.
 
-## Specification
-
-A Tango Property is a strict definition of a pair of key/value
-* configuration are represented in the form of properties.
-* data are represented in the form of attributes.
-* actions are represented in the form of commands.
-
-Its model can be represented as a defined tree which each elements are from a defined types: Class, Device, Property, Attribute, Command following the rules below:
+* To memorize the Attribute value each time it changes.
 
+## Specification
 
 A Tango Property is a strict definition of a pair of key/value
 * The Property SHALL have one key, called Property Name
-* The Property SHALL have one value
+* The Property SHALL have one value, which could be empty, called Property Value
+* If the Device is started upon a Tango Database, Class, Device, Attribute and Free Properties MAY be persisted into the Tango Database (RFC6)
+* If the Device is started without a Tango Database, the device and class properties MAY be persisted in a file (but no support for the attribute properties). 
+* All Device and Class Properties MAY be directly modified and accessed through the database device or its file (TODO: to be moved to Request/Reply RFC)
+* Attribute Properties SHOULD be modified and accessed at any time through the Device object (RFC-2) (TODO: to be moved to Request/Reply RFC)
+* A Property Value MAY have a default value
+* A Mandatory Property is a Property for which the Property Value SHALL NOT be null and SHALL not have a default value
+* If the Device is started upon a Tango Database, the Class and Device Properties changes are historized in the Tango Database (RFC-6) (TODO: To be moved to Tango Database RFC)
+* When a error occurs (i.e. network issue, Tango Database timeout...), a DevFailed exception MUST be thrown (RFC-x for DevFailed/Common types?) (TODO: To be moved to Request-Reply RFC)
+* All Device and Class Properties MAY be loaded at init command or device start-up (RFC-8)
+* A Device or Class Property MAY have a default value. 
+* The Device Property Value loading flow in pseudo-code is (TODO to be moved to RFC-8):
+	```
+	property-value = getDevicePropertyFromTangoDBorFile(device-name, property-name)
+	if(property-value is empty)
+  		property-value = getClassPropertyFromTangoDBorFile(class-name, property-name)
+	if(property-value is empty && default-value is not empty)
+  		property-value = default-value
+  	```
+	* The Class property value loading flow is:
+	```
+	property-value = getClassPropertyFromTangoDBorFile(class-name, property-name)
+	if(property-value is empty && default-value is not empty)
+  		property-value = default-value
+  	```
+* The precedence of Property Value for a given Property Name SHALL follow this order of preference:
+  1) the last Device Property Value
+  2) the last Class Property Value
+  3) the default Device Property Value
+  4) the default Class Property Value
```

Comment:
This is not completely true...
Here are the results of some tests I did:

| device prop. value in DB | class prop. value in DB | device prop. default value | device prop. mandatory in DB | class prop. default value | Prop value at the end of init_device |
| --- | --- | --- | --- | --- | --- |
| not defined | not defined | not defined | TRUE | not defined | Error |
| not defined | not defined | not defined | TRUE | Class Property Default Value | Class Property Default Value |
| Device Property Value in DB | not defined | not defined | TRUE | Class Property Default Value | Device Property Value in DB |
| Device Property Value in DB | not defined | not defined | TRUE | not defined | Device Property Value in DB |
| not defined | not defined | not defined | FALSE | not defined | empty string. 0.0 for double (not initialized in the code so, depends on compiler?) |
| not defined | not defined | not defined | FALSE | Class Property Default Value | Class Property Default Value |
| Device Property Value in DB | not defined | not defined | FALSE | Class Property Default Value | Device Property Value in DB |
| Device Property Value in DB | not defined | not defined | FALSE | not defined | Device Property Value in DB |
| not defined | not defined | Device Property Default Value | FALSE | not defined | Device Property Default Value |
| not defined | not defined | Device Property Default Value | FALSE | Class Property Default Value | Class Property Default Value |
| Device Property Value in DB | not defined | Device Property Default Value | FALSE | Class Property Default Value | Device Property Value in DB |
| Device Property Value in DB | not defined | Device Property Default Value | FALSE | not defined | Device Property Value in DB |
| not defined | Class Property Value in DB | not defined | TRUE | not defined | Class Property Value in DB |
| not defined | Class Property Value in DB | not defined | TRUE | Class Property Default Value | Class Property Value in DB |
| Device Property Value in DB | Class Property Value in DB | not defined | TRUE | Class Property Default Value | Device Property Value in DB |
| Device Property Value in DB | Class Property Value in DB | not defined | TRUE | not defined | Device Property Value in DB |
| not defined | Class Property Value in DB | not defined | FALSE | not defined | Class Property Value in DB |
| not defined | Class Property Value in DB | not defined | FALSE | Class Property Default Value | Class Property Value in DB |
| Device Property Value in DB | Class Property Value in DB | not defined | FALSE | Class Property Default Value | Device Property Value in DB |
| Device Property Value in DB | Class Property Value in DB | not defined | FALSE | not defined | Device Property Value in DB |
| not defined | Class Property Value in DB | Device Property Default Value | FALSE | not defined | Class Property Value in DB |
| not defined | Class Property Value in DB | Device Property Default Value | FALSE | Class Property Default Value | Class Property Value in DB |
| Device Property Value in DB | Class Property Value in DB | Device Property Default Value | FALSE | Class Property Default Value | Device Property Value in DB |
| Device Property Value in DB | Class Property Value in DB | Device Property Default Value | FALSE | not defined | Device Property Value in DB |

This is valid only for C++.
I found some of the results quite surprising. 
For instance, if you have a class property with a default value and a mandatory device property with the same name with no value at all in the DB, the device server will use the class default value in this case.
So, the property is defined as mandatory in DB but there is nothing in the DB.

When you have a class property with a default value and a device property with a default value, and nothing defined in the DB for these property, then the device server will use the class property default value.
With POGO, you cannot generate code

Reply:
Pogo gives an error when you try to generate code for a Java device server with a class property and a device property having the same property name.
For PythonHL, Pogo can generate the code in this use case, but the behaviour is different than in C++ since it is declaring 2 variables with the same name. The first one being the class property and the second one being the device property. So at the end, it is a device property.